src/persist-client/src/internal/cache.rs

```diff
@@ -118,16 +102,590 @@ impl Blob for BlobMemCache {
 
     async fn set(&self, key: &str, value: Bytes, atomic: Atomicity) -> Result<(), ExternalError> {
         let () = self.blob.set(key, value.clone(), atomic).await?;
-        self.cache
-            .insert(key.to_owned(), SegmentedBytes::from(value));
-        self.update_size_metrics();
+        let weight = value.len();
+        let mut cache = self.cache.lock().expect("lock poisoned");
+        cache.insert(key.to_owned(), SegmentedBytes::from(value), weight);
+        self.resize_and_update_size_metrics(&mut cache);
         Ok(())
     }
 
     async fn delete(&self, key: &str) -> Result<Option<usize>, ExternalError> {
         let res = self.blob.delete(key).await;
-        self.cache.invalidate(key);
-        self.update_size_metrics();
+        let mut cache = self.cache.lock().expect("lock poisoned");
+        cache.remove(key);
+        self.resize_and_update_size_metrics(&mut cache);
         res
     }
+
+    async fn restore(&self, key: &str) -> Result<(), ExternalError> {
+        self.blob.restore(key).await
+    }
+}
+
+mod lru {
+    use std::borrow::Borrow;
+    use std::hash::Hash;
+
+    use mz_ore::collections::HashMap;
+
+    /// A weighted cache, evicting the least recently used entries.
+    ///
+    /// This is reimplemented here, instead of using an existing crate, because
+    /// existing options seem to either not support weights or use unsafe.
+    pub struct Lru<K, V> {
+        evict_fn: Box<dyn Fn(K, V, usize) + Send>,
+        capacity: usize,
+        total_weight: usize,
+
+        nodes: List<LruNode<K, V>>,
+        by_key: HashMap<K, ListNodeId>,
```

Comment:
Haven't gone through the `Lru` code in detail yet, so maybe I'll figure this out, but I'm tempted to ask whether we could get away with something like:

```rust
nodes: HashMap<K, (V, Weight, Time)>,
by_time: BTreeMap<(Weight, K)>,
```

Since `BTreeMap` supports popping the smallest entry and also efficient removal.

Reply:
Not sure I follow! Is there maybe a typo or something in there? In particular, I don't really understand why the BTreeMap has a `Weight` key. If there's a way to avoid writing my own linked list, I'm quite interested to hear about it

Reply:
> Is there maybe a typo or something in there?

Sure was - fixed! Sorry about that - shouldn't have rushed it.

`BTreeSet` is there because we want to both add and remove arbitrary entries (as cache entries come and go) and to grab the minimum in an efficient way. Maybe a good 1:1 topic!

Reply:
This worked beautifully, thanks for the idea! I am very excited to not maintain an original doubly linked list implementation 😅

The one tweak I made was to replace the `BTreeSet<(Time, K)>` with a `BTreeMap<Time, K>`. This was 1) to avoid a couple unfortunate borrow checker issues (all workable but messy) and 2) to avoid an `Ord` requirement on `K` (also no biggie, but felt wrong conceptually). I don't really see any downside besides `Time`s needing to be globally unique, but they were anyway.

I went back and forth on whether the map should internally be ordered by increasing or decreasing time and ended up swapping the order compared to the list impl. I think I do prefer this one, but I verified that it's easy to slap a `std::cmp::Reverse` in there if you feel strongly the other way.